Mathlib/RingTheory/MvPowerSeries/LinearTopology.lean

```diff
@@ -0,0 +1,181 @@
+        convert zero_mem J
+        apply coeff_monomial_ne
+        rintro ⟨rfl⟩
+        exact h' (right_eq_inf.mpr he').symm
+      · specialize h e (le_refl e)
+        rwa [coeff_monomial_same] at h
+  · rintro ⟨hJK, hed⟩
+    exact basis_le hJK hed
+
+variable [TopologicalSpace R]
+
+-- We endow MvPowerSeries σ R with the product topology.
+open WithPiTopology
+
+/- variable (σ R) in
+theorem ringSubgroupsBasis :
+    RingSubgroupsBasis (fun (Jd : {J : TwoSidedIdeal R | (J : Set R) ∈ 𝓝 0} × (σ →₀ ℕ))
+        ↦ (basis σ R ⟨Jd.1, Jd.2⟩).asIdeal.toAddSubgroup) where
+  inter Jd Ke := ⟨⟨⟨Jd.1 ⊓ Ke.1, Filter.inter_mem Jd.1.prop Ke.1.prop⟩, Jd.2 ⊔ Ke.2⟩, by
+    simp only [le_inf_iff]
+    exact ⟨basis_le inf_le_left le_sup_left, basis_le inf_le_right le_sup_right⟩⟩
+  mul Jd := ⟨Jd, fun f ↦ by
+    simp only [Submodule.coe_toAddSubgroup, mem_mul]
+    rintro ⟨x, hx, y, hy, rfl⟩
+    exact Ideal.mul_mem_left _ _ hy⟩
+  leftMul f Jd := ⟨Jd, fun g hg ↦ (basis σ R ⟨Jd.1, Jd.2⟩).mul_mem_left f g hg⟩
+  rightMul f Jd := ⟨Jd, fun g hg ↦ by
+    intro e he
+    simp only [Submodule.coe_toAddSubgroup, TwoSidedIdeal.coe_asIdeal,
+      mem_coe, sub_zero, mem_basis_iff] at hg ⊢
+    classical
+    rw [coeff_mul]
+    apply sum_mem
+    rintro ⟨i, j⟩ h
+    apply TwoSidedIdeal.mul_mem_right
+    apply hg i (le_trans ?_ he)
+    simp only [← Finset.mem_antidiagonal.mp h, le_self_add]⟩
+-/
+
+/-- If the ring `R` is endowed with a linear topology, then the sets `↑basis σ R (J, d)`,
+for `J : TwoSidedIdeal R` which are neighborhoods of `0 : R` and `d : σ →₀ ℕ`,
+constitute a basis of neighborhoods of `0 : MvPowerSeries σ R` for the product topology. -/
+lemma hasBasis_nhds_zero [IsLinearTopology R R] [IsLinearTopology Rᵐᵒᵖ R] :
+    (𝓝 0 : Filter (MvPowerSeries σ R)).HasBasis
+      (fun Id : TwoSidedIdeal R × (σ →₀ ℕ) ↦ (Id.1 : Set R) ∈ 𝓝 0)
+      (fun Id ↦ basis _ _ Id) := by
+  rw [nhds_pi]
+  refine IsLinearTopology.hasBasis_twoSidedIdeal.pi_self.to_hasBasis ?_ ?_
+  · intro ⟨D, I⟩ ⟨hD, hI⟩
+    refine ⟨⟨I, Finset.sup hD.toFinset id⟩, hI, fun f hf d hd ↦ ?_⟩
+    rw [SetLike.mem_coe, mem_basis_iff] at hf
+    convert hf _ <| Finset.le_sup (hD.mem_toFinset.mpr hd)
+  · intro ⟨I, d⟩ hI
+    refine ⟨⟨Iic d, I⟩, ⟨finite_Iic d, hI⟩, ?_⟩
+    simpa [basis, coeff_apply, Iic, pi] using subset_rfl
+
+/-- The topology on `MvPowerSeries` is a left linear topology
+  when the ring of coefficients has a linar topology. -/
+instance [IsLinearTopology R R] [IsLinearTopology Rᵐᵒᵖ R] :
+    IsLinearTopology (MvPowerSeries σ R) (MvPowerSeries σ R) :=
+  IsLinearTopology.mk_of_hasBasis'  _ (hasBasis_nhds_zero) TwoSidedIdeal.mul_mem_left
+
+/-- The topology on `MvPowerSeries` is a right linear topology
+  when the ring of coefficients has a linear topology. -/
+instance [IsLinearTopology R R] [IsLinearTopology Rᵐᵒᵖ R] :
+    IsLinearTopology (MvPowerSeries σ R)ᵐᵒᵖ (MvPowerSeries σ R) :=
+  IsLinearTopology.mk_of_hasBasis'  _ (hasBasis_nhds_zero) TwoSidedIdeal.opp_smul_mem
```

Comment:
Can you use [TwoSidedIdeal.mul_mem_right](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/TwoSidedIdeal/Basic.html#TwoSidedIdeal.mul_mem_right) instead of the new lemma? Maybe you'd have to switch the order of the arguments.

Reply:
I could, this needs to write the full (fun _ _ _ _ _  mapsto ...). Is it better?

Reply:
I wouldn't say it's better, but it doesn't justify essentially duplicating the existing lemma.